File: apps/mobile_ios/test/routine_editor_build_test.dart
import 'package:flutter_test/flutter_test.dart';

import '../lib/routine_editor_build.dart';

({int? group, int? order}) _a(SupersetAssignment s) =>
    (group: s.supersetGroup, order: s.supersetOrder);

void main() {
  test('no flags → every exercise standalone', () {
    final out = assignSupersetGroups([false, false, false]).map(_a).toList();
    expect(out, [
      (group: null, order: null),
      (group: null, order: null),
      (group: null, order: null),
    ]);
  });

  test('single flag brackets a pair into one group, ordered 0 then 1', () {
    final out = assignSupersetGroups([true, false]).map(_a).toList();
    expect(out, [(group: 1, order: 0), (group: 1, order: 1)]);
  });

  test('a run of flags forms one longer group', () {
    final out = assignSupersetGroups([true, true, false]).map(_a).toList();
    expect(out, [
      (group: 1, order: 0),
      (group: 1, order: 1),
      (group: 1, order: 2),
    ]);
  });

  test('overlapping flags merge into one continuous group', () {
    final out =
        assignSupersetGroups([true, false, true, false]).map(_a).toList();
    expect(out, [
      (group: 1, order: 0),
      (group: 1, order: 1),
      (group: 1, order: 2),
      (group: 1, order: 3),
    ]);
  });

  test('two supersets split by a standalone exercise get distinct group ids',
      () {
    final out = assignSupersetGroups([true, false, false, false, true, false])
        .map(_a)
        .toList();
    expect(out, [
      (group: 1, order: 0),
      (group: 1, order: 1),
      (group: null, order: null),
      (group: null, order: null),
      (group: 2, order: 0),
      (group: 2, order: 1),
    ]);
  });

  test('a standalone exercise between two supersets stays ungrouped', () {
    final out = assignSupersetGroups([true, false, false, true, false])
        .map(_a)
        .toList();
    expect(out, [
      (group: 1, order: 0),
      (group: 1, order: 1),
      (group: null, order: null),
      (group: 2, order: 0),
      (group: 2, order: 1),
    ]);
  });

  test('a trailing flag is ignored (nothing follows to superset with)', () {
    final out = assignSupersetGroups([false, true]).map(_a).toList();
    expect(out, [(group: null, order: null), (group: null, order: null)]);
  });

  test('group + order are both-null or both-set (superset_chk invariant)', () {
    for (final a in assignSupersetGroups([true, false, false, true, true, false])) {
      expect(a.supersetGroup == null, a.supersetOrder == null);
    }
  });

  test('empty input yields no assignments', () {
    expect(assignSupersetGroups([]), isEmpty);
  });
}

File: apps/web/src/lib/gym/routine_editor_build.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';

import { assignSupersetGroups } from './routine_editor_build';

test('no flags → every exercise standalone', () => {
	const out = assignSupersetGroups([false, false, false]);
	assert.deepEqual(out, [
		{ supersetGroup: null, supersetOrder: null },
		{ supersetGroup: null, supersetOrder: null },
		{ supersetGroup: null, supersetOrder: null },
	]);
});

test('single flag brackets a pair into one group, ordered 0 then 1', () => {
	const out = assignSupersetGroups([true, false]);
	assert.deepEqual(out, [
		{ supersetGroup: 1, supersetOrder: 0 },
		{ supersetGroup: 1, supersetOrder: 1 },
	]);
});

test('a run of flags forms one longer group', () => {
	const out = assignSupersetGroups([true, true, false]);
	assert.deepEqual(out, [
		{ supersetGroup: 1, supersetOrder: 0 },
		{ supersetGroup: 1, supersetOrder: 1 },
		{ supersetGroup: 1, supersetOrder: 2 },
	]);
});

test('overlapping flags merge into one continuous group', () => {
	// ex0 supersets ex1, ex2 supersets ex3, but ex1↔ex2 are adjacent and ex2 is
	// flagged, so the run never breaks — all four are one group (a circuit).
	const out = assignSupersetGroups([true, false, true, false]);
	assert.deepEqual(out, [
		{ supersetGroup: 1, supersetOrder: 0 },
		{ supersetGroup: 1, supersetOrder: 1 },
		{ supersetGroup: 1, supersetOrder: 2 },
		{ supersetGroup: 1, supersetOrder: 3 },
	]);
});

test('two supersets split by a standalone exercise get distinct group ids', () => {
	const out = assignSupersetGroups([true, false, false, false, true, false]);
	assert.deepEqual(out, [
		{ supersetGroup: 1, supersetOrder: 0 },
		{ supersetGroup: 1, supersetOrder: 1 },
		{ supersetGroup: null, supersetOrder: null },
		{ supersetGroup: null, supersetOrder: null },
		{ supersetGroup: 2, supersetOrder: 0 },
		{ supersetGroup: 2, supersetOrder: 1 },
	]);
});

test('a standalone exercise between two supersets stays ungrouped', () => {
	const out = assignSupersetGroups([true, false, false, true, false]);
	assert.deepEqual(out, [
		{ supersetGroup: 1, supersetOrder: 0 },
		{ supersetGroup: 1, supersetOrder: 1 },
		{ supersetGroup: null, supersetOrder: null },
		{ supersetGroup: 2, supersetOrder: 0 },
		{ supersetGroup: 2, supersetOrder: 1 },
	]);
});

test('a trailing flag is ignored (nothing follows to superset with)', () => {
	const out = assignSupersetGroups([false, true]);
	assert.deepEqual(out, [
		{ supersetGroup: null, supersetOrder: null },
		{ supersetGroup: null, supersetOrder: null },
	]);
});

test('group + order are both-null or both-set (superset_chk invariant)', () => {
	for (const a of assignSupersetGroups([true, false, false, true, true, false])) {
		assert.equal(a.supersetGroup === null, a.supersetOrder === null);
	}
});

test('empty input yields no assignments', () => {
	assert.deepEqual(assignSupersetGroups([]), []);
});
